# Interpolação

<div class="alert alert-block alert-success">
    <b>Notas de aula baseadas no livro: </b> 
    <p>Algoritmos Numéricos, do autor Frederico Campos
</div>

A necessidade de obter um valor intermediário que não consta de uma tabela ocorre comumente. 
<img src="img/Picture31.jpg" width="50%" height="50%">

Dados experimentais, tabelas de estatísticas e de funções complexas são exemplos desta situação. 

Há diversas formas para realizar uma interpolação, a mais utilizada é a polinomial.

## Polinômios interpoladores
A forma mais simples para obter um polinômio interpolador, em termos conceituais, envolve a solução de um sistema de equações lineares. 

Será ilustrado como utilizar o polinômio quadrático. O caso linear e os de demais ordens são análogos.

### Interpolação quadrática
Sejam três pontos base $(x_0,y_0)$, $(x_1,y_1)$ e $(x_2,y_2)$, com $x_i$ distintos, de uma função $y=f(x)$. Para aproximar $f(z)$, $z\in (x_0,x_2)$, faz-se
$$f(x) \approx P_2(x)=a_o+a_1x + a_2x^2,$$
sendo $P_2(x)$ um polinômio interpolador de grau 2. Impondo que o polinômio interpolador passe pelos três pontos base, tem-se o sistema linear de ordem 3:

\begin{equation*}
\begin{cases}
\begin{array}{l}
P_2(x_0) = y_0\\
P_2(x_1) = y_1\\
P_2(x_2) = y_2\\
\end{array}
\end{cases}
\rightarrow
\begin{cases}
\begin{array}{c}
a_0 + a_1x_0 + a_2x_0^2=y_0\\
a_0 + a_1x_1 + a_2x_1^2=y_1\\
a_0 + a_1x_2 + a_2x_2^2=y_2\\
\end{array}
\end{cases}
\left[
\begin{array}{ccc}
1 & x_0 & x_0^2 \\
1 & x_1 & x_1^2 \\
1 & x_2 & x_2^2 \\
\end{array}
\right]
\left[
\begin{array}{c}
a_0\\
a_1\\
a_2\\
\end{array}
\right]
=
\left[
\begin{array}{c}
y_0\\
y_1\\
y_2\\
\end{array}
\right]
\end{equation*}

A matriz dos coeficientes é a matriz de Vandermonde e o sistema de equações admite solução única, pois o determinante dessa matriz é não nulo. 

Consequentemente, por três pontos distintos passa um único polinômio de grau menor ou igual a dois. 

Este fato pode ser generalizado dizendo que  por $n+1$ pontos distintos passa um único polinômio de grau menor ou igual a $n$.

**Exemplo:** Calcular $P_2(0.2)$ usando os dados da tabela abaixo:
\begin{array}{|c|c|c|c|}
	\hline x & 0.1 & 0.6 & 0.8 \\ 
	\hline y & 1.221 & 3.320 & 4.953 \\ 
	\hline 
\end{array} 

**Solução:**
\begin{equation}
\left[
\begin{array}{ccc}
1 & 0.1 & 0.01 \\
1 & 0.6 & 0.36 \\
1 & 0.8 & 0.64 
\end{array}
\right]
\left[
\begin{array}{c}
a_0\\
a_1\\
a_2\\
\end{array}
\right]
=
\left[
\begin{array}{c}
1.221\\
3.320\\
4.953\\
\end{array}
\right]
\implies
a=
\left[
\begin{array}{r}
1.141\\
0.231\\
5.667\\
\end{array}
\right]
\implies
P_2(x)=1.141+0.231x+5.667x^2
\end{equation}

Assim, pode ser verificado que $P_2(0.2)=1.414$. 

**Exercicio:** Resolva o sistema linear acima usando a função LU para encontrar o vetor de coeficientes $a$ e utilize a função polyval com o vetor de coeficientes $a$ para verificar que $P_2(x)$ passa pelos pontos base e que $P_2(0.2)=1.414$. Compare os coeficientes com aqueles encontrados pela função polyfit.



In [21]:
x = [0.1 0.6 0.8];
y = [1.221 3.320 4.953];
a = polyfit(x, y, 2)
polyval(a, 0.2)

a =

   5.66714   0.23100   1.14123

ans =  1.4141


No entanto, [matrizes de Vandermonde](https://pt.wikipedia.org/wiki/Matriz_de_Vandermonde) são muito mal condicionadas; ou seja, suas soluções são muito sensíveis a erros de arredondamento. 

**Exercicio:** Monte o sistema linear para os três ultimos pontos da tabela acima e utilize a função **cond()** para calcular o número de condicionamento para a matriz de coeficientes.

A =

   1.000000   0.100000   0.010000
   1.000000   0.600000   0.360000
   1.000000   0.800000   0.640000

ans =  34.410


Essa metodologia, apesar de ser conceitualmente simples, está sujeita a imprecisão e demanda certo esforço computacional, visto que a decomposição $LU$ ou a Eliminição de Gauss tem complexidade da ordem de $n^3$. 

Existem outras formas de obter o polinômio interpolador com menor esforço, a saber, os métodos de Lagrange e de Newton.

## Polinômios de Lagrange
Sejam dados $n+1$ pontos $(x_0,y_0)$,$(x_1,y_1)$,$\dots$,$(x_n,y_n)$, sendo $x_i$ distintos, tais que 
$y_i=f(x_i)$ e $x \in (x_0,x_n)$. 

Deseja-se construir um polinômio $L_n(x)$ com grau menor ou igual a $n$ que passe por todos os $n+1$ pontos listados.

Isto é, $L_n(x_i)=y_i$, $i=0,1,2,\dots,n$

Inicialmente, sejam os polinômios de grau $n$ $P_i(x)$, $i=0,1,2,\dots,n$, que satisfazem duas condições básicas: 
\begin{equation}
\begin{cases}
\begin{array}{l}
P_i(x_i)\ne 0\\
P_i(x_j)=0, \forall i \ne j
\end{array}
\end{cases}
\end{equation}

Assim, cada $P_i(x)$ pode ser descrito como:

\begin{array}{l}
P_0(x) = (x-x_1)(x-x_2)\cdots(x-x_n)\\[0.5cm]
P_1(x) = (x-x_0)(x-x_2)\cdots(x-x_n)\\[0.5cm]
\vdots\\
P_n(x) = (x-x_0)(x-x_1)\cdots(x-x_{n-1})\\
\end{array}

Ou seja, 
$$P_i(x)=\prod_{j=0, j \ne i}^{n}(x - x_j), i=0,1,2,...,n$$

O polinômio $L_n(x)$ pode ser escrito como uma combinação linear dos polinômios $P_i(x)$:
$$L_n(x)=c_0P_0(x) + c_1P_1(x)+c_2P_2(x)+\dots+c_nP_n(x)$$   

$$L_n(x)=\sum_{i=0}^{n}c_iP_i(x)$$

Tendo em vista que $P_i(x_j)=0, \forall i \ne j$ e que $L_n(x_i)=y_i$, $i=0,1,2,\dots,n$

$$L_n(x_i)=y_i=c_iP_i(x_i) \leftarrow c_i=\frac{y_i}{P_i(x_i)}$$

Substituindo o valor de $c_i$ em $L_n(x)$ temos:

$$L_n(x)=\sum_{i=0}^{n}\frac{y_i}{P_i(x_i)}P_i(x)$$

Por fim, a fórmula do polinomio interpolador de Lagrange de grau $n$ é:

$$L_n(x)=\sum_{i=0}^{n}y_i\prod_{j=0, j \ne i}^{n}\frac{x-x_j}{x_i-x_j}$$

**Exemplo:** Calcular $L_2(0.2)$ do caso anterior utilizando a forma de Lagrange.

**Solução:**

É preciso calcular $L_2(0.2)=y_0PP_0(0.2) + y_1PP_1(0.2) + y_2PP_2(0.2)$, onde $PP_i(0.2), i=0,1,2$ é dado por:

\begin{array}{l}
PP_0(0.2)=\dfrac{(x-x_1)(x-x_2)}{(x_0-x_1)(x_0-x_2)}=\dfrac{(0.2-0.6)(0.2-0.8)}{(0.1-0.6)(0.1-0.8)}=0.6857\\[0.5cm]
PP_1(0.2)=\dfrac{(x-x_0)(x-x_2)}{(x_1-x_0)(x_1-x_2)}=\dfrac{(0.2-0.1)(0.2-0.8)}{(0.6-0.1)(0.6-0.8)}=0.6\\[0.5cm]
PP_2(0.2)=\dfrac{(x-x_0)(x-x_1)}{(x_2-x_0)(x_2-x_1)}=\dfrac{(0.2-0.1)(0.2-0.6)}{(0.8-0.1)(0.8-0.6)}=-0.2857\\
\end{array}

E finalmente, 
$$L_2(0.2)=1.221(0.6857) + 3.320(0.6) + 4.953(-0.2857) \approx 1.414.$$

Obviamente, a interpolação de Lagrange requer um menor esforço computacional do que aquela que necessita resolver um sistema de equações lineares.

### Algoritmo e Complexidade
\begin{array}[l]
~1 . {\tt Algoritmo~Polinômio\_de\_Lagrange}\\
2 . \{{\tt Parâmetros~de~entrada:} n, x, y, z\}\\
(número~de~pontos,~x~e~y~da~tabela~e~z~o~valor~a~ser~interpolado)\\
3 . \{{\tt Parâmetros~de~saída:} p~(resultado~da~interpolação)\}\\
4 . \quad p~\leftarrow~0\\
5 . \quad {\tt para}~i~\leftarrow~1~{\tt até}~n+1~{\tt faça}\\
6 . \qquad c~\leftarrow~1;~d~\leftarrow~1\\
7 . \qquad {\tt para}~j~\leftarrow~1~{\tt até}~n+1~{\tt faça}\\
8 . \qquad\qquad if i \ne j\\
9 . \qquad\qquad\qquad c~\leftarrow~c * (z-x(j));~d~\leftarrow~d * (x(i)-x(j))\\
10. \qquad {\tt fim-para}\\
11. \qquad p~\leftarrow~p + y(i)*c/d\\
12. \quad {\tt fim-para}\\
13. {\tt fim-algoritmo}
\end{array}

Nota-se facilmente que o algoritmo é da ordem de $n^2$, sendo $n$ o grau do polinômio interpolador. 

**Exercício:** Implemente o algoritmo de interpolação pelo método de Lagrange e compare o resultado e o tempo de execução com o método de interpolação quadrática.


## Polinômio de Newton

### Operador de diferença divida
Seja a função $y=f(x)$ que passa pelos pontos $(x_i,y_i)$, $i=0,1,2,\dots,n$. O operador de diferença dividida $\Delta$ é definido como sendo 

a) ordem 0: $\Delta^0 y_i=y_i=f[x_i]$,

b) ordem 1: $\Delta^1 y_i = \dfrac{\Delta^0 y_{i+1}-\Delta^0 y_i}{x_{i+1}-x_i}=\dfrac{y_{i+1}-y_i}{x_{i+1}-x_i}=f[x_i,x_{i+1}]$,

c) ordem 2: $\Delta^2 y_i = \dfrac{\Delta^1 y_{i+1}-\Delta^1 y_i}{x_{i+2}-x_i}=f[x_i,x_{i+1},x_{i+2}]$,

d) ordem $n$: $\Delta^n y_i = \dfrac{\Delta^{n-1} y_{i+1}-\Delta^{n-1} y_i}{x_{i+n}-x_i}=f[x_i,x_{i+1},\dots,x_{i+n}]$.

**Teorema 1** - Se $y=f(x)$ for um polinômio de grau $n$, então suas diferenças divididas de ordem $n+1$ são identicamente nulas, isto é, $$f[x_0,x_1,\dots,x_n]=0 ~\forall x.$$

**Exemplo:** Verificar a tabela de diferenças divididas de $y=5x^3-2x^2-x+3$ para alguns pontos $x_i$ no intervalo [0;0.9].

\begin{array}{|c|c|c|c|c|c|c|}
	\hline i & x_i & y_i & \Delta^1 y_i & \Delta^2 y_i  & \Delta^3 y_i & \Delta^4 y_i \\ 
	\hline\hline 0 & 0.0 & 3.000 &  &  &  &  \\ 
	\hline & & & -1.20 & & &\\
	\hline 1 & 0.2 & 2.760 &  & 0.5 &  &  \\ 
	\hline & & & -1.05 & & 5 & \\
	\hline 2 & 0.3 & 2.655 &  & 2.5 &  & 0 \\ 
	\hline & & & -0.55 & & 5 & \\
	\hline 3 & 0.4 & 2.600 &  & 5.0 &   & 0 \\ 
	\hline & & & 1.45& & 5  & \\
	\hline 4 & 0.7 & 3.035 &   &   8.0  &   &  \\ 
	\hline & & & 5.45 & & & \\
	\hline 5 & 0.9 & 4.125 &       &     &   &  \\ 
	\hline 
\end{array} 

### Fórmula de Newton
Sejam os $n+1$ pontos $(x_i,y_i)$, $i=0,1,2,\dots,n$, com $x_i$ distintos, tais que $y_i=P(x_i)$, sendo $P(x)$ um polinômio de grau $n$. Pela definição de diferenças divididas tem-se:
$$f[x,x_o]=\dfrac{P(x)-P(x_0)}{x-x_0}$$
ou
$$P(x) = P(x_0) + f[x,x_0](x-x_0)$$
No entanto, a diferença dividida de ordem 2:
$$f[x,x_0,x_1]=\dfrac{f[x,x_0]-f[x_0,x_1]}{x-x_1}\rightarrow f[x,x_0]=f[x_0,x_1] + f[x,x_0,x_1](x-x_1).$$
fazendo a substituição do termo,
$$P(x)=P(x_0) + f[x_0,x_1](x-x_0) + f[x,x_0,x_1](x-x_0)(x-x_1).$$
De forma indutiva é possível inferir que 
\begin{equation*}
\begin{array}{ll}
P(x) &= P(x_0) + f[x_0,x_1](x-x_0) + f[x_0,x_1,x_2](x-x_0)(x-x_1)+\\[0.2cm]
&+f[x_0,x_1,x_2,x_3](x-x_0)(x-x_1)(x-x_2)+\dots+\\[0.2cm] 
&+f[x_0,x_1,\dots,x_n](x-x_0)(x-x_1)\dots(x-x_{n-1})+\\[0.2cm] 
&+f[x,x_0,x_1,\dots,x_n](x-x_0)(x-x_1)\dots(x-x_n).\\
&\end{array}
\end{equation*}
Considerando que $P(x)$ é um polinômio de grau $n$, então pelo Teorema 1, resulta que 

$$f[x,x_0,x_1,\dots,x_n]=0$$

Fazendo $y_0=P(x_0)$ e usando a notação de diferenças divididas com o operador $\Delta$, tem-se o polinômio de Newton de grau $n$ 
$$P_n(x)=y_0+\Delta y_0(x-x_0)+\Delta^2y_0(x-x_0)(x-x_1)+\dots+\Delta^n y_0(x-x_0)\dots(x-x_{n-1})$$

$$P_n(x) = y_0 + \sum_{i=1}^{n}\Delta^i y_0 \prod_{j=0}^{i-1}(x-x_j)$$

Uma vantagem dos polinômios de Newton sobre os de Lagrange é que para aumentar o grau basta acrescentar um novo termo em vez de montar o polinômio novamente.

**Exemplo:** Dada a tabela de diferenças divididas já calculada abaixo determine:

\begin{array}{|c|c|c|c|c|c|c|}
    \hline i & x_i & y_i & \Delta y_i & \Delta^2 y_i  & \Delta^3 y_i & \Delta^4 y_i \\ 
    \hline
    \hline 0 & 0.1 & 0.31620 &  1.15750 & -0.93167  & 0.64667 &  0.42222 \\ 
    \hline 1 & 0.3 & 0.54770 &  0.87800 & -0.60833  & 0.90000 &  \\ 
    \hline 2 & 0.4 & 0.63550 &  0.69550 & -0.24833  &  &  \\ 
    \hline 3 & 0.6 & 0.77460 &  0.62100 &  &   &  \\ 
    \hline 4 & 0.7 & 0.83670 &   &   &   &  \\ 
    \hline 
\end{array} 

a) $P_2(0.55)$.

b) $P_4(0.2)$.

**Solução:**

a) Sendo $P_2(x)=y_0+\Delta y_0(x-x_0)+\Delta^2y_0(x-x_0)(x-x_1)$, como o ponto de interpolação tem o grau menor que o total de pontos da tabela, deve ser feita uma escolha. Assim, $x_0=0.4, x_1=0.6, x_2=0.7$. 

E portanto,
$$P_2(0.55)=0.63550+0.69550(0.55-0.4)+(-0.24833)(0.55-0.4)(0.55-0.6)=0.74169$$

b) Como o grau do polinômio é 4, a tabela será utilizada por completo. Assim,
\begin{equation*}
\begin{array}{ll}
P_4(x) & = y_0 + \Delta y (x-x_0) + \Delta y^2(x-x_0)(x-x_1)
       +\Delta y^3 (x-x_0)(x-x_1)(x-x_2)+\\[0.2cm]  
       & \Delta y^4(x-x_0)(x-x_1)(x-x_2)(x-x_3)
\end{array}
\end{equation*}

E portanto,
\begin{equation*}
\begin{array}{ll}
P_4(0.2) & = 0.3162 + 1.1575 (0.1) + (-0.93167)(0.1)(-0.1)
+0.64667(0.1)(-0.1)(-0.2)+\\[0.2cm]  
& 0.42222(0.1)(-0.1)(-0.2)(-0.4)=0.44222
\end{array}
\end{equation*}



In [6]:
x = [0 0.2 0.3 0.4 0.7 0.9]';
y = [3 2.76 2.655 2.6 3.035 4.125]';
f = @(x) 5.*x.^3 -2.*x.^2 -x + 3;
xx = 0.5;

addpath('src/')
[yy, b] = newtint(x, y, xx)

e = f(xx)-yy

yy =  2.6250
b =

   3.00000  -1.20000   0.50000   5.00000  -0.00000   0.00000
   2.76000  -1.05000   2.50000   5.00000   0.00000   0.00000
   2.65500  -0.55000   5.00000   5.00000   0.00000   0.00000
   2.60000   1.45000   8.00000   0.00000   0.00000   0.00000
   3.03500   5.45000   0.00000   0.00000   0.00000   0.00000
   4.12500   0.00000   0.00000   0.00000   0.00000   0.00000

e =   -4.4409e-16


In [1]:
x = [0.1 0.3 0.4 0.6 0.7]';
y = [0.3162 0.5477 0.6355 0.7746 0.8367]';

addpath('src/')
[p4, b] = newtint(x, y, 0.2)

p4 =  0.44222
b =

   0.31620   1.15750  -0.93167   0.64667   0.42222
   0.54770   0.87800  -0.60833   0.90000   0.00000
   0.63550   0.69550  -0.24833   0.00000   0.00000
   0.77460   0.62100   0.00000   0.00000   0.00000
   0.83670   0.00000   0.00000   0.00000   0.00000



## Erro de truncamento da interpolação polinomial
Existe um erro ao aproximar uma $f(x)$ por um polinômio $P_n(x)$. Que é denominado erro de truncamento e definido por 
$$T_n(x)=\dfrac{f^{n+1}(\xi)}{(n+1)!}(x-x_0)(x-x_1)\dots(x-x_n), ~x_0<\xi<x_n.$$

Sendo $f(x)$ definida no intervalo $[a,b]$ que contém os pontos $x_0,x_1,\dots,x_n$, e supondo que a derivada $f^{n+1}(x)$ exista e seja contínua no intervalo $(a,b)$. Na prática encontrar $\xi$ não é uma tarefa simples portanto utilizamos a nossa de erro máximo:
$$|T_n(x)|\leq \dfrac{M_{n+1}}{(n+1)!}|(x-x_0)(x-x_1)\dots(x-x_n)|.$$

onde $M_{n+1}=max|f^{n+1}(x)|,~x\in[a,b]$.


**Exemplo:** Seja a tabela de pontos:
\begin{array}{|c|c|c|c|c|c|c|c|}
	\hline x & 1.1 & 1.4 & 1.9 & 2.1 & 2.5 & 3.0 & 3.2 \\ 
	\hline f(x) & 0.6942 & 0.6952 & 1.1759 & 1.6562 & 3.4325 & 8.0855 & 11.0925  \\ 
	\hline 
\end{array} 

Determine:
- O valor de $P_2(2.2)$
- Determinar o erro máximo como metido sabendo que $f^{'''}(x)=e^x$.

**Solução:**

a) Pode ser utilizado Lagrange ou Newton. Assim, utilizando uma tabela de diferenças divididas com os pontos $x_0=1.9$, $x_1=2.1$ e $x_2=2.5$.
\begin{array}{|c|c|c|c|c|}
    \hline i & x_i & y_i & \Delta y_i & \Delta^2 y_i \\ 
    \hline\hline 0 & 1.9 & 1.1759 & 2.4015 & 3.3988 \\ 
    \hline       1 & 2.1 & 1.6562 & 4.4408 &  \\ 
    \hline       2 & 2.5 & 3.4325 &  & \\ 
    \hline
\end{array} 

com $P_2(2.2)=1.1759+2.4015(2.2-1.9)+3.3988(2.2-1.9)(2.2-2.1)=1.9983.$


b) Como $T_2(x) \leq \dfrac{M_3}{3!}(x-x_0)(x-x_1)(x-x_2)$ e $M_3  = \underset{x \in [x_0,x_2]}{max|f'''(x)|} = e^{2.5}$ 

tem-se que uma cota superior para o erro seria
$$|T_2(2.2)| \leq \dfrac{e^{2.5}}{3!}|(2.2-1.9)(2.2-2.1)(2.2-2.5)|=0.0183.$$
e portanto,
$$|f(2.2)-P_2(2.2)|=|1.9850-1.9983|=0.0133<0.0183.$$ 



### Estimativa do erro cometido utilizando diferenças divididas
O Teorema 1 garante as diferenças divididas de ordem $n+1$ são iguais a zero caso $f(x)$ seja um polinômio de grau $n$. De forma geral, as diferenças divididas servem como um parâmetro para estimar o erro cometido na interpolação polinomial. Sendo assim,
$$|T_n(x)| \approx \mathrm{max}|\mathrm{diferenca~dividida~de~ordem}~n+1|\times|(x-x_0)(x-x1) \dots (x-x_n)|$$

** Exemplo:** Estime o erro cometido ao calcular $P_2(0.55 )$ supondo que  
não exista nenhuma outra informação de $f(x)$ além dos pontos tabelados.
\begin{array}{|c|c|c|c|c|c|c|}
    \hline i & x_i & y_i & \Delta y_i & \Delta^2 y_i  & \Delta^3 y_i & \Delta^4 y_i \\ 
    \hline\hline 0 & 0.1 & 0.3162 & 1.1575 & -1.0317 & 1.1468 & -1.2447 \\ 
    \hline 1 & 0.3 & 0.5477 & 0.8480 & -0.4583 & 0.4000 &  \\ 
    \hline 2 & 0.4 & 0.6355 & 0.7105 & -0.2983 &  &  \\ 
    \hline 3 & 0.6 & 0.7746 & 0.6210 &  &   &  \\ 
    \hline 4 & 0.7 & 0.8367 &   &     &   &  \\ 
    \hline 
\end{array} 

**Solução:**
Como $|T_2(x)| \approx \mathrm{max}|\mathrm{diferenca~dividida~de~ordem}~3|\times|(x-x_0)(x-x_1)(x-x_2)|$, tem-se que
$$|T_2(0.55)| \approx |1.1468|\times|(0.55-0.4)(0.55-0.6)(0.55-0.7)|=1.2902 \times 10^{-3}.$$



## Interpolação Inversa


